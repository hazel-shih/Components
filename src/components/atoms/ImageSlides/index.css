:root {
	--current-image: var(--current-image);
}

.image-slides {
	display: flex;
	width: 100%;
	height: 100vh;
	justify-content: center;
	align-items: center;
	position: relative;
	overflow: hidden;

	& > .black-background {
		position: absolute;
		width: 100%;
		height: 100vh;
		background-image: var(--current-image);
		background-position: center center;
		background-size: cover;
		filter: brightness(40%) blur(2px);
		transform: scale(1.1);
		transition: 0.4s ease;
	}

	& > .current-image-container {
		position: relative;
		font-size: 50px;
		width: 70%;
		height: 70vh;
		background-position: center center;
		background-size: cover;
		background-image: var(--current-image);
		transition: 0.4s ease;

		& > .arrow {
			position: absolute;
			cursor: pointer;
			z-index: 100;
			transition: 0.3s ease;

			&.left {
				left: -50px;
				top: 50%;
				transform: translateY(-50%);

				&:hover {
					transform: translateY(-50%) translateX(-10px);
				}
			}
			&.right {
				right: -50px;
				top: 50%;
				transform: translateY(-50%);

				&:hover {
					transform: translateY(-50%) translateX(10px);
				}
			}
			& > path {
				fill: white;
			}
		}
	}
}
