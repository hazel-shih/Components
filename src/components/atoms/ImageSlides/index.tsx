import React, { useState, CSSProperties } from 'react';
import classnames from 'classnames';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCaretLeft, faCaretRight } from '@fortawesome/free-solid-svg-icons';

import styles from './index.css';

type ImageList = {
	src: string;
}[];

interface ImageSlidesProperty {
	imageList: ImageList;
	className?: string;
}

const getImageUrl = (imageSrc: string) => `url(${imageSrc})`;

const ImageSlides: React.FC<ImageSlidesProperty> = ({ imageList, className }) => {
	const [currentImageIndex, setCurrentImageIndex] = useState<number>(0);
	return (
		<div className={classnames(styles.imageSlides, className)}>
			<div
				className={styles.blackBackground}
				style={
					{ '--current-image': getImageUrl(imageList[currentImageIndex].src) } as CSSProperties
				}
			/>
			<div
				className={styles.currentImageContainer}
				style={
					{ '--current-image': getImageUrl(imageList[currentImageIndex].src) } as CSSProperties
				}
			>
				<FontAwesomeIcon
					icon={faCaretLeft}
					className={classnames(styles.arrow, styles.left)}
					onClick={() => {
						if (currentImageIndex === 0) {
							setCurrentImageIndex(imageList.length - 1);
						} else {
							setCurrentImageIndex(currentImageIndex - 1);
						}
					}}
				/>
				<FontAwesomeIcon
					icon={faCaretRight}
					className={classnames(styles.arrow, styles.right)}
					onClick={() => {
						if (currentImageIndex === imageList.length - 1) {
							setCurrentImageIndex(0);
						} else {
							setCurrentImageIndex(currentImageIndex + 1);
						}
					}}
				/>
			</div>
		</div>
	);
};

export default ImageSlides;
