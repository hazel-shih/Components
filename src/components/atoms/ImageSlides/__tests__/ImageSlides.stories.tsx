import React from 'react';
import { storiesOf } from '@storybook/react';

import ImageSlides from 'components/atoms/ImageSlides';

const stories = storiesOf('atoms/ImageSlides', module);

const testImageList = [
	{
		src: 'https://images.unsplash.com/photo-1473186578172-c141e6798cf4?crop=entropy&cs=tinysrgb&fm=jpg&ixlib=rb-1.2.1&q=80&raw_url=true&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1073',
	},
	{
		src: 'https://images.unsplash.com/photo-1521170813716-0b3f42fcfb65?ixlib=rb-1.2.1&raw_url=true&q=80&fm=jpg&crop=entropy&cs=tinysrgb&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1170',
	},
	{
		src: 'https://images.unsplash.com/photo-1475924156734-496f6cac6ec1?crop=entropy&cs=tinysrgb&fm=jpg&ixlib=rb-1.2.1&q=80&raw_url=true&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1170',
	},
	{
		src: 'https://images.unsplash.com/photo-1505228395891-9a51e7e86bf6?ixlib=rb-1.2.1&raw_url=true&q=80&fm=jpg&crop=entropy&cs=tinysrgb&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1633',
	},
	{
		src: 'https://images.unsplash.com/photo-1490365728022-deae76380607?ixlib=rb-1.2.1&raw_url=true&q=80&fm=jpg&crop=entropy&cs=tinysrgb&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1074',
	},
];

stories.add('__interactive', () => <ImageSlides imageList={testImageList} />);
