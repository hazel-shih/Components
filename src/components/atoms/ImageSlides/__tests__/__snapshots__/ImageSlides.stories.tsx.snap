// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`__interactive 1`] = `
<DocumentFragment>
  <div
    class="imageSlides"
  >
    <div
      class="blackBackground"
      style="--current-image: url(https://images.unsplash.com/photo-1473186578172-c141e6798cf4?crop=entropy&cs=tinysrgb&fm=jpg&ixlib=rb-1.2.1&q=80&raw_url=true&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1073);"
    />
    <div
      class="currentImageContainer"
      style="--current-image: url(https://images.unsplash.com/photo-1473186578172-c141e6798cf4?crop=entropy&cs=tinysrgb&fm=jpg&ixlib=rb-1.2.1&q=80&raw_url=true&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1073);"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-caret-left arrow left"
        data-icon="caret-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 256 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M137.4 406.6l-128-127.1C3.125 272.4 0 264.2 0 255.1s3.125-16.38 9.375-22.63l128-127.1c9.156-9.156 22.91-11.9 34.88-6.943S192 115.1 192 128v255.1c0 12.94-7.781 24.62-19.75 29.58S146.5 415.8 137.4 406.6z"
          fill="currentColor"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-caret-right arrow right"
        data-icon="caret-right"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 256 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M118.6 105.4l128 127.1C252.9 239.6 256 247.8 256 255.1s-3.125 16.38-9.375 22.63l-128 127.1c-9.156 9.156-22.91 11.9-34.88 6.943S64 396.9 64 383.1V128c0-12.94 7.781-24.62 19.75-29.58S109.5 96.23 118.6 105.4z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
